### ISSUE TITLE
Consider allowing Additional Fields to be persisted real-time.

### ISSUE THREAD

**Issue by senadir** [woothemes/woocommerce#42079]
Right now, we will introduce 2 types of fields, ones in the address sections, and ones in the contact and additional info section.

Fields in Address will persist real time, meaning they will be sent to the server the moment you blur them or after a short delay. Fields in other sections won't, and will be persisted once you place the order.

This leaves a problem because certain fields might need to be persisted in real-time, for example a gift wrap checkbox might add a fee to the order, so it needs to be persisted in real time.

Persisting in real time means sending a request to Store API and getting an updated **Cart** back. This is a bit tricky because fields are to likely be stored on the order meta but we still need to get a cart back to see any meaningful UI changes since the Checkout response doesn't help create the Checkout view.

**Comment by senadir** [comment]
We discussed this issue in our call with @mikejolley and @opr 

The idea was we would probably need to support `PUT /checkout` request that would update Checkout, we're still ensure about requests that would cause the cart totals to be updated, so we're likely to return cart as well if needed.

**Comment by mikejolley** [comment]
With my latest PRs, persisting works after "place order", but not during form input. That would require support on the update-customer endpoint, or we could refactor checkout to PUT to checkout endpoint perhaps.

**Comment by webdados** [comment]
Yes please!

> This leaves a problem because certain fields might need to be persisted in real-time, for example a gift wrap checkbox might add a fee to the order, so it needs to be persisted in real time.

Exactly my current issue. I'm adding a select field in which each option can have an associated fee value. If this field is created on the address section it works fine, but if I register it on the contact or additional information sections (which is the logical place to put it) it doesn't persist.

**Comment by webdados** [comment]
Here's my feedback after some exchanging some ideas with @opr and @senadir on Slack.

I'm trying to develop the setting of order fees based on the value of custom fields (address or non-address).

This is possible if the field is an address one, as it's value is passed along all the API requests made on the checkout when any field (address or email) is manipulated by the user, so that I can use those values on the `woocommerce_cart_calculate_fees` filter. They're also stored on the order draft.

If the custom fields are non-address, this doesn't work bacause:

- There's no API request when they're changed
- They're not saved into the order draft until the order is finished
- They're not included on any API request
- The values set by the customer are lost if the checkout page is reloaded

This makes it impossible to implement such a feature and I guess many other features that rely on custom fields values, so IMHO these fields should be treated like all the other fields: be available on all the API requests, trigger API requests on changing values, be stored on the order draft on any change, in short: be persistent.

I know that currently the address and email fields are set to be persistent as they're part of the "address" set of fields. Custom fields outside the address are not, and the additional info ones are not even saved to the customer, and that's probably the reason they we're not designed to be persistent, but keep in mind that in a non-customized checkout all the existing fields are persistent. It's not important for the customer or shop owner what kind of fields they are, neither it should matter. If a shop owner uses the API to add a custom field (address or not), he expects it to function like all the other fields. The customer will also have a bad experience if he refreshed the checkout page, or goes back to the store to add another product and then when coming back to the checkout, the data he entered to those custom fields is lost, but no the data on the core fields.

In short: this is a developer problem (because a lot of stuff cannot be build on top os custom fields) but also an end-user UX problem.

**Comment by senadir** [comment]
> but keep in mind that in a non-customized checkout all the existing fields are persistent.

Not really, terms and conditions, order note, and selected payment method are not persisted, and we kinda used the same system, which is why we're in this pickle.

The issue for now is that address and coupons exists on Cart, but all other fields exists on Checkout, and we have no system to keep that up to date  in a Checkout lifecycle (we only update the Checkout on place order).

Now we can add a PUT to Checkout to make it updated during Checkout, and this would solve the issue of persisting fields just fine, but this won't solve the issue with fees for example, in which it needs an updated cart.

**Comment by webdados** [comment]
> Not really, terms and conditions, order note, and selected payment method are not persisted, and we kinda used the same system, which is why we're in this pickle.

Right, my bad.

> Now we can add a PUT to Checkout to make it updated during Checkout, and this would solve the issue of persisting fields just fine, but this won't solve the issue with fees for example, in which it needs an updated cart.

I see. That's a bummer. It's not urgent for me, I was just playing around trying to implement this thing mostly for fun and to learn more about the "new" checkout, but it's still an issue that makes it very hard to implement anything that touches the cart based on custom fields.

I've added onChange events to the non-address fields to trigger an `extensionCartUpdate` and then realized it would work just on that request and everything is lost after any other interaction with the checkout. I could save those values into the session and mimic the persistent status of the fields, but I guess it would be too much "hammering" and complex to maintain.

**Comment by webdados** [comment]
🎉
